### PR TITLE
assert: consider failed condition unreachable with NDEBUG

### PR DESCRIPTION
**Pull request by benpicco**
When `NDEBUG` is set, asserts are usually removed from the code. But we can do better: With debugging enabled, the `__assert_func()` that is called on the failed condition is marked as noreturn, so the compiler optimizes the remaining code as if the failed condition would never be taken.

By marking the failed condition as `__assert_func()` the compiler is able to do the same optimization with `NDEBUG` set.

### CODE REVIEW

File: newlib/libc/include/assert.h [hunk 1]
```diff
@@ -43,8 +43,8 @@ extern "C" {
 
 #undef assert
 
-#ifdef NDEBUG           /* required by ANSI standard */
-# define assert(__e) ((void)0)
+#ifdef NDEBUG
+# define assert(__e) ((__e) ? (void)0 : __unreachable())
```
Comment (rdiez) [inline newlib/libc/include/assert.h:47]: I do not think this is right. In release builds (when NDEBUG is defined), the assert code should be removed, so the original code is correct.

Also, why did you remove comment "required by ANSI standard". Is the comment wrong?

**Comment by jnohlgard** [comment]
The description of this PR is not very clear on the rationale on why we don't want to remove the assert code for NDEBUG. Is there a clear benefit to keep the code and let the c compiler optimize it away as unreachable compared to stripping it in the preprocessor stage?

File: newlib/libc/include/assert.h (hunk 1 above)
Comment (benpicco) [reply newlib/libc/include/assert.h:47]: Well, this slightly diverges from the ANSI standard now.
But I can see how this might have some undesirable consequences, maybe a separate `assume()` is a better solution here.

Comment (keith-packard) [reply newlib/libc/include/assert.h:47]: I can see benefits to leaving the code visible to the compiler so that it gets validated. However, the C standard is prescriptive about the assert macro:

> the assert macro is defined simply as
> ```
> #define assert(ignore) ((void)0)
> ```

I don't think we can vary from that and still remain standards compliant.

**Comment by keith-packard** [comment]
> The description of this PR is not very clear on the rationale on why we don't want to remove the assert code for NDEBUG. Is there a clear benefit to keep the code and let the c compiler optimize it away as unreachable compared to stripping it in the preprocessor stage?

Yup, I'd love to have the assert code validated to make sure changes don't invalidate it. Even better, I'd love to know that turning on NDEBUG would not introduce changes in behavior, but I suspect that would require additional language changes.

Maybe someday?

**Comment by jnohlgard** [comment]
Perhaps switch to C++? 😅

**Comment by rdiez** [comment]
I actually develop embedded software in C++, and I still find NDEBUG a good compromise. 

Of course turning it on will change some behaviour: the code will run more slowly, and something extra may have to be initialised beforehand, like support for printing the failed assert message on the UART debug console.
